donkeycar/parts/lidar.py

```diff
@@ -218,6 +401,114 @@ def run(self, distances, angles):
     def shutdown(self):
         pass
 
+class LidarPlot2(object):
+    '''
+    takes the raw lidar measurements and plots it to an image
+    '''
+    PLOT_TYPE_LINE = 0
+    PLOT_TYPE_CIRC = 1
+    def __init__(self, resolution=(500,500),
+                 max_dist=1000, #mm
+                 radius_plot=3,
+                 plot_type=PLOT_TYPE_CIRC,
+                 rotate_plot=0,
+                 angle_direction=COUNTER_CLOCKWISE):  # direction of increasing angles in the data
+        self.frame = Image.new('RGB', resolution)
+        self.max_dist = max_dist
+        self.rad = radius_plot
+        self.resolution = resolution
+        if plot_type == self.PLOT_TYPE_CIRC:
+            self.plot_fn = self.plot_circ
+        else:
+            self.plot_fn = self.plot_line
+        self.angle_direction = angle_direction
+        self.rotate_plot = rotate_plot
+
+    def plot_line(self, img, dist, theta, max_dist, draw):
+        '''
+        scale dist so that max_dist is edge of img (mm)
+        and img is PIL Image, draw the line using the draw ImageDraw object
+        '''
+        if dist < 0 or dist > max_dist:
+            return  # don't print out of range pixels
+
+        cx = img.width / 2
+        cy = img.height / 2
+        max_pixel = min(cx, cy)
+        dist = dist / max_dist * max_pixel
+        
+        theta = (theta + self.rotate_plot) % 360.0
+            
+        if self.angle_direction != COUNTER_CLOCKWISE:
+            theta = (360.0 - (theta % 360.0)) % 360.0
+            
+        theta = np.radians(theta)
+        sx = cx + math.cos(theta) * dist
+        sy = cy - math.sin(theta) * dist
+        ex = cx + math.cos(theta) * (dist + self.rad)
+        ey = cy - math.sin(theta) * (dist + self.rad)
+        fill = 128
+        draw.line((sx,sy, ex, ey), fill=(fill, fill, fill), width=1)
+        
+    def plot_circ(self, img, dist, theta, max_dist, draw):
+        '''
+        scale dist so that max_dist is edge of img (mm)
+        and img is PIL Image, draw the circle using the draw ImageDraw object
+        '''
+        if dist < 0 or dist > max_dist:
+            return  # don't print out of range pixels
+
+        cx = img.width / 2
+        cy = img.height / 2
+        max_pixel = min(cx, cy)
+        dist = dist / max_dist * max_pixel
+
+        theta = (theta + self.rotate_plot) % 360.0
+
+        if self.angle_direction != COUNTER_CLOCKWISE:
+            theta = (360.0 - (theta % 360.0)) % 360.0
+            
+        theta = np.radians(theta)
+        sx = int(cx + math.cos(theta) * dist)
+        sy = int(cy - math.sin(theta) * dist)
+        ex = int(cx + math.cos(theta) * (dist + 2 * self.rad))
+        ey = int(cy - math.sin(theta) * (dist + 2 * self.rad))
+        fill = 128
```

Comment:
This looks pretty much like a copy of line 435ff - can you factor out that common code in order to reduce duplication / redundancy?

Reply:
I started doing this (by breaking out the draw methods into functions that the class methods would call) but I found that the resulting methods would be tightly coupled and fragile because the part treat we treat coordinates and angles and how they output data differently.  I don't think these two lidar parts will coexist forever; I think we should deprecate LidarPlot in favour of LidarPlot2, but I don't want to remove it right away and break anyones code.  I would like to give them time to move to the newer part before we remove the older part and I would like to get the new part field tested.  So there is not much upside to trying to share code when we will just delete one anyway and there is  downside because it makes the code coupled and complex.  However, the thought process did get me going and I refactored the drawing code so it is stand-alone.  That is most of the code anyway.  So if we wanted to we could make the older RPLidar part use it (again, not worth the effort if we are going to remove it in the future).   We may want to move the drawing code into its own file.

Reply:
Thanks for explaining, that makes a lot of sense. Maybe put a one-line comment into the code to explain that the old part is expected to be deprecated.